src/FsCodec.SystemTextJson/UnionConverter.fs

```diff
@@ -25,3 +29,111 @@ module private Union =
+           && (typedefof<Option<_>> = t.GetGenericTypeDefinition()
+                || t.GetGenericTypeDefinition().IsValueType)) // Nullable<T>
+
+    let typeHasJsonConverterAttribute = memoize (fun (t : Type) -> t.IsDefined(typeof<Serialization.JsonConverterAttribute>, false))
+
+    let propTypeRequiresConstruction (propertyType : Type) =
+        not (isInlinedIntoUnionItem propertyType)
+        && not (typeHasJsonConverterAttribute propertyType)
+
+    /// Prepare arguments for the Case class ctor based on the kind of case and how F# maps that to a Type
+    /// and/or whether we need to let json.net step in to convert argument types
+    let mapTargetCaseArgs (element : JsonElement) options : PropertyInfo[] -> obj [] = function
+        | [| singleCaseArg |] when propTypeRequiresConstruction singleCaseArg.PropertyType ->
+            [| JsonSerializer.DeserializeElement (element, singleCaseArg.PropertyType, options) |]
+        | multipleFieldsInCustomCaseType ->
+            [| for fi in multipleFieldsInCustomCaseType ->
+                match element.GetProperty fi.Name with
+                // TOTHINK: I'm not sure this applies with STJ
+
+                //| null when
+                //    // Afford converters an opportunity to handle the missing field in the best way I can figure out to signal that
+                //    // The specific need being covered (see tests) is to ensure that, even with MissingMemberHandling=Ignore,
+                //    // the TypeSafeEnumConverter should reject missing values
+                //    // not having this case would go direct to `null` without passing go
+                //    typeHasJsonConverterAttribute fi.PropertyType
+                //    || serializer.MissingMemberHandling = MissingMemberHandling.Error ->
+                //        // NB caller can opt out of erroring by setting NullValueHandling = NullValueHandling.Ignore)
+                //        // which renders the following equivalent to the next case
+                //        JToken.Parse("null").ToObject(fi.PropertyType, serializer)
+                | el when el.ValueKind = JsonValueKind.Null -> null
+                | el -> JsonSerializer.DeserializeElement (el, fi.PropertyType, options) |]
+
+/// Serializes a discriminated union case with a single field that is a
+/// record by flattening the record fields to the same level as the discriminator
+type UnionConverter<'T> private (discriminator : string, ?catchAllCase) =
+    inherit Serialization.JsonConverter<'T>()
+
+    new() = UnionConverter("case")
+    new(discriminator: string, catchAllCase: string) = UnionConverter(discriminator, ?catchAllCase = match catchAllCase with null -> None | x -> Some x)
+
+    override __.CanConvert (t : Type) = Union.tryGetUnion t |> Option.isSome
+
+    override __.Write(writer, value, options) =
+        let value = box value
+        let union = Union.tryGetUnion (value.GetType()) |> Option.get // TOASK: Do we wanna keep the try?
```

Comment:
I know `Option.get` is abhorred, but [`_.CanConvert(Type)` does guard us](https://github.com/jet/FsCodec/pull/43/files#diff-3a3077af061d875164a9a2412e664557R80)

Reply:
Yes, I tend to break 'the law' (if I'm feeling guilty, I'd consider writing `// CanConvert guarantees the tryGetUnion will yield a result`. Alternately, one can argue that anyone that's on .NET writing a Converter with a CanConvert 3 lines above it doesn't need the noise of stating the obvious at them!)